### PR TITLE
Fixed #32479 -- Added fallbacks to subsequent language codes in translations.

### PR DESCRIPTION
**Pull request by Cup0fCoffee**
Added zh-Hans-CN, zh-Hans-HK, zh-Hans-MO, zh-Hans-SG, zh-Hans-TW
languages with fallback to zh-hans.
Added zh-Hant-CN, zh-Hant-HK, zh-Hant-MO, zh-Hant-SG, zh-Hant-TW
languages with fallback to zh-hant.

### CODE REVIEW

File: django/conf/locale/__init__.py [hunk 1]
```diff
@@ -587,12 +587,42 @@
...
+    },
+    'zh-Hans-MO': {
+        'fallback': ['zh-hans'],
+    },
+    'zh-Hans-SG': {
+        'fallback': ['zh-hans'],
+    },
+    'zh-Hans-TW': {
+        'fallback': ['zh-hans'],
+    },
```
Comment (felixxm) [inline django/conf/locale/__init__.py:604]: IMO `hans`/`hant` should be lowercased, e.g. `zh-hans-SG`.

File: django/conf/locale/__init__.py [hunk 2]
```diff
@@ -587,12 +587,42 @@
...
+    'zh-Hant-MO': {
+        'fallback': ['zh-hant'],
+    },
+    'zh-Hant-SG': {
+        'fallback': ['zh-hant'],
+    },
+    'zh-Hant-TW': {
+        'fallback': ['zh-hant'],
+    },
```
Comment (felixxm) [inline django/conf/locale/__init__.py:625]: Please add new fallabacks to the `tests.i18n.tests.MiscTests.test_special_fallback_language()` (use `self.subTest()`).

**Review by felixxm — COMMENTED** [review]
@Cup0fCoffee Thanks for this patch :+1: 

@claudep Can you take a look?

File: django/conf/locale/__init__.py (hunk 1 above)
Comment (claudep) [reply django/conf/locale/__init__.py:604]: In the IANA registry, they are written as `Hans`, even if the codes should be handled case insensitively.

https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry

**Comment by claudep** [comment]
I'm wondering if we should not doing such fallbacks automatically, when the first part of the code is known.

Typically in `django.utils.translation.get_supported_language_variant` . See the comment `# If 'fr-ca' is not supported, try special fallback or language-only 'fr'.` So it would be an extension of this logic.

**Comment by felixxm** [comment]
> I'm wondering if we should not doing such fallbacks automatically, when the first part of the code is known.
> 
> Typically in `django.utils.translation.get_supported_language_variant` . See the comment `# If 'fr-ca' is not supported, try special fallback or language-only 'fr'.` So it would be an extension of this logic.

Sounds good, so if 'zh-Hant-TW' is not supported try `'zh-Hant`, if `'zh-Hant` is not supported try `zh`, and so on.

**Comment by Cup0fCoffee** [comment]
> I'm wondering if we should not doing such fallbacks automatically, when the first part of the code is known.
> 
> Typically in `django.utils.translation.get_supported_language_variant` . See the comment `# If 'fr-ca' is not supported, try special fallback or language-only 'fr'.` So it would be an extension of this logic.

Do you mean something like this:
```diff
diff --git a/django/utils/translation/trans_real.py b/django/utils/translation/trans_real.py
index 8042f6fdc4..b29ebf86b2 100644
--- a/django/utils/translation/trans_real.py
+++ b/django/utils/translation/trans_real.py
@@ -475,13 +475,15 @@ def get_supported_language_variant(lang_code, strict=False):
     """
     if lang_code:
         # If 'fr-ca' is not supported, try special fallback or language-only 'fr'.
-        possible_lang_codes = [lang_code]
+        possible_lang_codes = []
         try:
             possible_lang_codes.extend(LANG_INFO[lang_code]['fallback'])
         except KeyError:
             pass
-        generic_lang_code = lang_code.split('-')[0]
-        possible_lang_codes.append(generic_lang_code)
+        lang_code_split = lang_code.split('-')
+        generic_lang_code = lang_code_split[0]
+        for i in range(len(lang_code_split), 0, -1):
+            possible_lang_codes.append('-'.join(lang_code_split[:i]))
         supported_lang_codes = get_languages()
 
         for code in possible_lang_codes:
```

I have a feeling, that there is a more elegant way to write it, I just want to confirm that I understood the logic you implied correctly.

**Comment by claudep** [comment]
Yes, I think you got the idea.

File: django/conf/locale/__init__.py (hunk 2 above)
Comment (Cup0fCoffee) [reply django/conf/locale/__init__.py:625]: Is this ok? Also, should it contain tests with lower case language codes?
```diff
diff --git a/tests/i18n/tests.py b/tests/i18n/tests.py
index 8248122a97..69c7cd968f 100644
--- a/tests/i18n/tests.py
+++ b/tests/i18n/tests.py
@@ -1419,10 +1419,25 @@ class MiscTests(SimpleTestCase):
         Some languages may have special fallbacks that don't follow the simple
         'fr-ca' -> 'fr' logic (notably Chinese codes).
         """
-        r = self.rf.get('/')
-        r.COOKIES = {}
-        r.META = {'HTTP_ACCEPT_LANGUAGE': 'zh-my,en'}
-        self.assertEqual(get_language_from_request(r), 'zh-hans')
+        tests = [
+            ('zh-my', 'zh-hans'),
+            ('zh-Hans-CN', 'zh-hans'),
+            ('zh-Hans-HK', 'zh-hans'),
+            ('zh-Hans-MO', 'zh-hans'),
+            ('zh-Hans-SG', 'zh-hans'),
+            ('zh-Hans-TW', 'zh-hans'),
+            ('zh-Hant-CN', 'zh-hant'),
+            ('zh-Hant-HK', 'zh-hant'),
+            ('zh-Hant-MO', 'zh-hant'),
+            ('zh-Hant-SG', 'zh-hant'),
+            ('zh-Hant-TW', 'zh-hant'),
+        ]
+        for value, expected in tests:
+            with self.subTest(value=value):
+                r = self.rf.get('/')
+                r.COOKIES = {}
+                r.META = {'HTTP_ACCEPT_LANGUAGE': f'{value},en'}
+                self.assertEqual(get_language_from_request(r), expected)
 
     def test_parse_language_cookie(self):
         """
```

Comment (claudep) [reply django/conf/locale/__init__.py:625]: Good-looking, but no need to test each variant, one for Hans and one for Hant should be enough. And yes, I would also add different cases to test case insensibility.

File: tests/i18n/tests.py [hunk 3]
```diff
@@ -1419,10 +1419,23 @@ def test_special_fallback_language(self):
...
+        tests = [
+            ('zh-my', 'zh-hans'),
+            # canonical
+            ('zh-Hans-CN', 'zh-hans'),
+            ('zh-Hant-HK', 'zh-hant'),
+            # lowercase variations
+            ('zh-hans-mo', 'zh-hans'),
+            ('zh-hant-tw', 'zh-hant'),
+            ('zh-hans-HK', 'zh-hans'),
+            ('zh-hant-SG', 'zh-hant'),
```
Comment (felixxm) [inline tests/i18n/tests.py:1431]: These are not related with the `fallback` feature so I moved them to a separate test.

**Review by felixxm — COMMENTED** [review]
@Cup0fCoffee Thanks for updates :+1:  Welcome aboard :boat: 

I squashed commits and pushed small edits.

File: django/utils/translation/trans_real.py [hunk 4]
```diff
@@ -474,14 +474,17 @@ def get_supported_language_variant(lang_code, strict=False):
...
         possible_lang_codes = [lang_code]
         try:
             possible_lang_codes.extend(LANG_INFO[lang_code]['fallback'])
         except KeyError:
             pass
-        generic_lang_code = lang_code.split('-')[0]
-        possible_lang_codes.append(generic_lang_code)
+        lang_code_split = lang_code.split('-')
+        for i in range(len(lang_code_split) - 1, 0, -1):
+            possible_lang_codes.append('-'.join(lang_code_split[:i]))
```
Comment (pope1ni) [inline django/utils/translation/trans_real.py:486]: While longer, this avoids creating the extra list, string building and complex range calculations:

```suggestion
        i = None
        try:
            while i := lang_code.rindex('-', 0, i):
                possible_lang_codes.append(lang_code[:i])
        except ValueError:
            pass
```

I know it also uses the walrus operator, but Django 4.0 is targeting Python 3.8+, so it is available to us. It seems much more readable to me.

(If this isn't a performance critical path then `contextlib.suppress()` could be used to shave off two lines.)

Comment (felixxm) [reply django/utils/translation/trans_real.py:486]: @pope1ni Thanks for the suggestion, but IMO it's less readable and it's more complex (as far as I'm aware) :thinking: :
<details>

```pycon
>>> import dis
>>> def test_1():
...     possible_lang_codes = []
...     lang_code = 'zh-hans-mo'
...     lang_code_split = lang_code.split('-')
...     for i in range(len(lang_code_split) - 1, 0, -1):
...         possible_lang_codes.append('-'.join(lang_code_split[:i]))
... 
>>> def test_2():
...     possible_lang_codes = []
...     lang_code = 'zh-hans-mo'
...     i = None
...     try:
...         while i := lang_code.rindex('-', 0, i):
...             possible_lang_codes.append(lang_code[:i])
...     except ValueError:
...         pass
... 
>>> dis.dis(test_1)
  2           0 BUILD_LIST               0
              2 STORE_FAST               0 (possible_lang_codes)

  3           4 LOAD_CONST               1 ('zh-hans-mo')
              6 STORE_FAST               1 (lang_code)

  4           8 LOAD_FAST                1 (lang_code)
             10 LOAD_METHOD              0 (split)
             12 LOAD_CONST               2 ('-')
             14 CALL_METHOD              1
             16 STORE_FAST               2 (lang_code_split)

  5          18 LOAD_GLOBAL              1 (range)
             20 LOAD_GLOBAL              2 (len)
             22 LOAD_FAST                2 (lang_code_split)
             24 CALL_FUNCTION            1
             26 LOAD_CONST               3 (1)
             28 BINARY_SUBTRACT
             30 LOAD_CONST               4 (0)
             32 LOAD_CONST               5 (-1)
             34 CALL_FUNCTION            3
             36 GET_ITER
        >>   38 FOR_ITER                28 (to 68)
             40 STORE_FAST               3 (i)

  6          42 LOAD_FAST                0 (possible_lang_codes)
             44 LOAD_METHOD              3 (append)
             46 LOAD_CONST               2 ('-')
             48 LOAD_METHOD              4 (join)
             50 LOAD_FAST                2 (lang_code_split)
             52 LOAD_CONST               0 (None)
             54 LOAD_FAST                3 (i)
             56 BUILD_SLICE              2
             58 BINARY_SUBSCR
             60 CALL_METHOD              1
             62 CALL_METHOD              1
             64 POP_TOP
             66 JUMP_ABSOLUTE           38
        >>   68 LOAD_CONST               0 (None)
             70 RETURN_VALUE
>>> dis.dis(test_2)
  2           0 BUILD_LIST               0
              2 STORE_FAST               0 (possible_lang_codes)

  3           4 LOAD_CONST               1 ('zh-hans-mo')
              6 STORE_FAST               1 (lang_code)

  4           8 LOAD_CONST               0 (None)
             10 STORE_FAST               2 (i)

  5          12 SETUP_FINALLY           42 (to 56)

  6     >>   14 LOAD_FAST                1 (lang_code)
             16 LOAD_METHOD              0 (rindex)
             18 LOAD_CONST               2 ('-')
             20 LOAD_CONST               3 (0)
             22 LOAD_FAST                2 (i)
             24 CALL_METHOD              3
             26 DUP_TOP
             28 STORE_FAST               2 (i)
             30 POP_JUMP_IF_FALSE       52

  7          32 LOAD_FAST                0 (possible_lang_codes)
             34 LOAD_METHOD              1 (append)
             36 LOAD_FAST                1 (lang_code)
             38 LOAD_CONST               0 (None)
             40 LOAD_FAST                2 (i)
             42 BUILD_SLICE              2
             44 BINARY_SUBSCR
             46 CALL_METHOD              1
             48 POP_TOP
             50 JUMP_ABSOLUTE           14
        >>   52 POP_BLOCK
             54 JUMP_FORWARD            20 (to 76)

  8     >>   56 DUP_TOP
             58 LOAD_GLOBAL              2 (ValueError)
             60 COMPARE_OP              10 (exception match)
             62 POP_JUMP_IF_FALSE       74
             64 POP_TOP
             66 POP_TOP
             68 POP_TOP

  9          70 POP_EXCEPT
             72 JUMP_FORWARD             2 (to 76)
        >>   74 END_FINALLY
        >>   76 LOAD_CONST               0 (None)
             78 RETURN_VALUE
```

</details>

Comment (pope1ni) [reply django/utils/translation/trans_real.py:486]: @felixxm Silly me... I forgot that `str.rfind()` exists which returns `-1` when not found whereas `str.rindex()` raises `ValueError`. That is actually what I wanted all along.

```suggestion
        i = None
        while (i := lang_code.rfind('-', 0, i)) >= 0:
            possible_lang_codes.append(lang_code[:i])
```

The exception handling added lots of overhead. This is much better than both the versions above if we're going to consider `dis.dis()`...
<details>

```pycon
>>> import dis
>>> def test_3():
...     possible_lang_codes = []
...     lang_code = 'zh-hans-mo'
...     i = None
...     while (i := lang_code.rfind('-', 0, i)) >= 0:
...         possible_lang_codes.append(lang_code[:i])
... 
>>> dis.dis(test_3)
  2           0 BUILD_LIST               0
              2 STORE_FAST               0 (possible_lang_codes)

  3           4 LOAD_CONST               1 ('zh-hans-mo')
              6 STORE_FAST               1 (lang_code)

  4           8 LOAD_CONST               0 (None)
             10 STORE_FAST               2 (i)

  5     >>   12 LOAD_FAST                1 (lang_code)
             14 LOAD_METHOD              0 (rfind)
             16 LOAD_CONST               2 ('-')
             18 LOAD_CONST               3 (0)
             20 LOAD_FAST                2 (i)
             22 CALL_METHOD              3
             24 DUP_TOP
             26 STORE_FAST               2 (i)
             28 LOAD_CONST               3 (0)
             30 COMPARE_OP               5 (>=)
             32 POP_JUMP_IF_FALSE       54

  6          34 LOAD_FAST                0 (possible_lang_codes)
             36 LOAD_METHOD              1 (append)
             38 LOAD_FAST                1 (lang_code)
             40 LOAD_CONST               0 (None)
             42 LOAD_FAST                2 (i)
             44 BUILD_SLICE              2
             46 BINARY_SUBSCR
             48 CALL_METHOD              1
             50 POP_TOP
             52 JUMP_ABSOLUTE           12
        >>   54 LOAD_CONST               0 (None)
             56 RETURN_VALUE
```

</details>

But I was more interested in the readability of it and think this is better. (Although what probably made it harder to understand for me was the long variable name - `lang_code_split` - I always find that the longer it is, the less clear it is.)

Anyway, even if you still reject this, I just thought it was a nice use of `:=` and the `stop` argument of `str.rfind()`. Python has many gems like this that are overlooked. 😄

Comment (Cup0fCoffee) [reply django/utils/translation/trans_real.py:486]: @pope1ni IMO, it's an elegant solution and a nice use of the walrus operator.
If walrus operator is a problem (I like it in this case, but it's a controversial feature, as it breaks backward comparability, and the resulted code seems less 'pythonic' and more C-like, which might be harder to read for some people), you can rewrite it as:
```suggestion
        i = len(lang_code)
        while i > -1:
            i = lang_code.rfind('-', 0, i)
            possible_lang_codes.append(lang_code[:i])
```
Even in this state, it's better than what I originally wrote, because it doesn't create a new list and doesn't use `join`, which should be better for readability (no long variable names), uses less memory and should, in theory, perform better.

Comment (pope1ni) [reply django/utils/translation/trans_real.py:486]: @Cup0fCoffee Thanks. Your version is also good if we're rejecting mine because it has the walrus operator.

> If walrus operator is a problem (I like it in this case, but it's a controversial feature, as it breaks backward compatibility, and the resulted code seems less 'pythonic' and more C-like, which might be harder to read for some people)

Ah, yes. I never understand this argument of things being "unpythonic" because a feature may have come from another language. Python, like many languages, is evolving all the time and it makes sense to take good ideas from other places. The same negative argument is playing out for addition of pattern matching. Typically those making these claims are used to the language as it is and don't want to see it change or have another thing to learn. Often they are teaching beginners and think these changes are too complex to teach while disregarding those who are experienced and have a need for these features. At the end of the day people can use which language features they like and ignore those they don't if they're anti.

I would love to see [PEP 505](https://www.python.org/dev/peps/pep-0505/) come to fruition. Granted it is more syntactical sugar, but it allows for writing much more concise code in some cases. The argument against is that it doesn't read as clearly to some, but then many layers of nested blocks are less clear too, at least in my opinion. Readability is also subjective, a matter of perception - something that is readable to "you" might not be as readable to me.

In the case of `:=`, the token is uniquely used for assignment expressions and cannot be confused with `=` as would be possible in C. The backward compatibility bit is moot as we're Python 3.8+ only on master. I can't see how else it would be backward incompatible.

(Please don't take this as me implying you're for or against in any way... Was just stating my viewpoint on the subject.)

Comment (felixxm) [reply django/utils/translation/trans_real.py]: Thanks both :+1: I pushed edits.

**Comment by felixxm** [comment]
Test failure is not related with this patch.
